### PR TITLE
P1207R4 Movability of Single-pass Iterators

### LINKED ISSUE #3026: [2019-07 LWG Motion 22] P1207R4 Movability of Single-pass Iterators

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #3026.

### CODE REVIEW

**Comment by burblebee** [comment]
To lwg: In change to [counted.iter.access], should "\returns" become "\effects Equivalent to:" to mimic change in P0896R4 (e85af9533)?

**Comment by jwakely** [comment]
> To lwg: In change to [counted.iter.access], should "\returns" become "\effects Equivalent to:" to mimic change in P0896R4 (e85af95)?

I don't see any great need to do that here. Also I see that [is.sorted] p4 is wrong since that commit, it's missing a `return`.

**Comment by jwakely** [comment]
To be clear: it's certainly not the case that all _Returns:_ elements should be replaced by "_Effects:_ Equivalent to:". They mean different things.

**Comment by burblebee** [comment]
> These changes seem to be P1207R4 not P1638R1.

Yup.  I didn't know how to fix that without force-pushing, hopefully that was ok.

File: source/iterators.tex [hunk 1]
```diff
@@ -4043,14 +4051,30 @@
 
 \indexlibrarymember{base}{move_iterator}%
 \begin{itemdecl}
-constexpr Iterator base() const;
+constexpr Iterator base() const &;
 \end{itemdecl}
 
 \begin{itemdescr}
+\pnum
+\constraints \tcode{Iterator} satisfies \tcode{CopyConstructible}.
```
Comment (jwakely) [inline source/iterators.tex]: I think this should be `\libconcept{CopyConstructible}`.

File: source/iterators.tex [hunk 2]
```diff
@@ -4043,14 +4051,30 @@
...
+
+\pnum
+\effects \tcode{Iterator} models \tcode{CopyConstructible}.
```
Comment (jwakely) [inline source/iterators.tex]: Same here.

**Comment by burblebee** [comment]
> ... I see that [is.sorted] p4 is wrong since that commit, it's missing a `return`.

Don't think we should address that as part of this commit - please open a new issue?

File: source/iterators.tex [hunk 3]
```diff
@@ -4945,7 +4969,8 @@
       requires Assignable<I&, const I2&>
         constexpr counted_iterator& operator=(const counted_iterator<I2>& x);
 
-    constexpr I base() const;
+    constexpr I base() const & requires CopyConstructible<I>;;
```
Comment (jwakely) [inline source/iterators.tex]: Two semi-colons at the end of this line.

**Comment by jwakely** [comment]
> > ... I see that [is.sorted] p4 is wrong since that commit, it's missing a `return`.
> 
> Don't think we should address that as part of this commit - please open a new issue?

Sure, I'm fixing it separately.

File: source/iterators.tex (hunk 1 above)
Comment (burblebee) [reply source/iterators.tex]: Ok, changed to "the \oldconcept{CopyConstructible} requirements".

File: source/iterators.tex (hunk 2 above)
Comment (burblebee) [reply source/iterators.tex]: ditto.

File: source/iterators.tex (hunk 3 above)
Comment (burblebee) [reply source/iterators.tex]: thanks!

File: source/iterators.tex [hunk 4]
```diff
@@ -4043,14 +4051,32 @@
...
 \indexlibrarymember{base}{move_iterator}%
 \begin{itemdecl}
-constexpr Iterator base() const;
+constexpr Iterator base() const &;
 \end{itemdecl}
 
 \begin{itemdescr}
+\pnum
+\constraints \tcode{Iterator} satisfies
+the \oldconcept{CopyConstructible} requirements.
```
Comment (jwakely) [inline source/iterators.tex]: Should be `\libconcept` not `\oldconcept`, here and in the next paragraph.

Comment (jwakely) [reply source/iterators.tex]: Use \oldconcept for "meets the _Cpp17Foo_ requirements" and use \libconcept for "models `Bar`" and "satisfies `Baz`".

Comment (CaseyCarter) [reply source/iterators.tex]: More precisely, this should be `satisfies \libconcept{CopyConstructible}.` without "the" and "requirements" - we only use "the XXXX requirements" for the old named requirement sets, not with concepts.

File: source/iterators.tex [hunk 5]
```diff
@@ -4043,14 +4051,32 @@
...
+
+\pnum
+\effects \tcode{Iterator} models
+the \oldconcept{CopyConstructible} requirements.
```
Comment (CaseyCarter) [inline source/iterators.tex]: Ditto `models \libconcept{CopyConstructible}.`

Comment (CaseyCarter) [reply source/iterators.tex]: Oh, and `\effects` should be `\expects` here.

File: source/iterators.tex [hunk 6]
```diff
@@ -5088,14 +5115,24 @@
...
 \end{itemdescr}
 
+\indexlibrarymember{base}{counted_iterator}%
+\begin{itemdecl}
+constexpr I base() &&;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns \tcode{std::move(current)}.
```
Comment (CaseyCarter) [inline source/iterators.tex]: Bah, the ingress paper had incorrect context for line 5123 so we didn't notice the "*Effects*: Equivalent to:  `return current;`" / "*Returns*: `std::move(current)`." inconsistency.  I don't suppose I could convince the Editors to reword this paragraph to "*Effects*: Equivalent to: `return std::move(current);`" for consistency to save the trouble of an LWG issue?

Comment (burblebee) [reply source/iterators.tex]: I'm afraid that's a normative change, but it's clear that that was what was intended.  @zygoloid what do you think?

File: source/iterators.tex (hunk 4 above)
Comment (burblebee) [reply source/iterators.tex]: Yes, the oldconcept was wrong here (my bad), but thanks to both of you for explaining the difference in wording between the two - hope I'll remember it for next time!

Fixed.

File: source/iterators.tex (hunk 5 above)
Comment (burblebee) [reply source/iterators.tex]: thanks!

File: source/iterators.tex (hunk 6 above)
Comment (burblebee) [reply source/iterators.tex]: @CaseyCarter Please open a separate issue for this (might have to be a LWG issue); I'm not comfortable making that change as part of this motion.  Thanks!

**Pull request merged**